File: e2e-tests/utils/test-helpers.ts
/**
 * Test helper utilities
 */

import { Page, expect } from "@playwright/test";

/**
 * Wait for navigation to complete
 */
export async function waitForNavigation(page: Page, url?: string) {
  if (url) {
    await page.waitForURL(url);
  } else {
    await page.waitForLoadState("networkidle");
  }
}

/**
 * Wait for API response
 */
export async function waitForAPIResponse(
  page: Page,
  urlPattern: string | RegExp,
) {
  return page.waitForResponse((response) => {
    const url = response.url();
    if (typeof urlPattern === "string") {
      return url.includes(urlPattern);
    }
    return urlPattern.test(url);
  });
}

/**
 * Take a screenshot with automatic naming
 */
export async function takeScreenshot(page: Page, name: string) {
  await page.screenshot({
    path: `screenshots/${name}-${Date.now()}.png`,
    fullPage: true,
  });
}

/**
 * Check if element is visible
 */
export async function isVisible(
  page: Page,
  selector: string,
): Promise<boolean> {
  try {
    const element = page.locator(selector);
    return await element.isVisible({ timeout: 5000 });
  } catch {
    return false;
  }
}

/**
 * Wait for text to appear
 */
export async function waitForText(page: Page, text: string, timeout = 10000) {
  await page.waitForSelector(`text=${text}`, { timeout });
}

/**
 * Fill form field by label
 */
export async function fillByLabel(page: Page, label: string, value: string) {
  const input = page.locator(
    `label:has-text("${label}") + input, label:has-text("${label}") input`,
  );
  await input.fill(value);
}

/**
 * Click button by text
 */
export async function clickButton(page: Page, text: string) {
  await page.click(`button:has-text("${text}")`);
}

/**
 * Assert toast/notification appears
 */
export async function expectToast(page: Page, message: string) {
  const toast = page.locator(
    `[role="alert"]:has-text("${message}"), .toast:has-text("${message}")`,
  );
  await expect(toast).toBeVisible({ timeout: 5000 });
}

/**
 * Assert loading state completes
 */
export async function waitForLoadingComplete(page: Page) {
  // Wait for any loading spinners or skeletons to disappear
  await page
    .waitForSelector('.loading, [data-loading="true"], .skeleton', {
      state: "hidden",
      timeout: 10000,
    })
    .catch(() => {
      // If no loading indicator found, that's fine
    });
}

/**
 * Get text content of element
 */
export async function getTextContent(
  page: Page,
  selector: string,
): Promise<string> {
  const element = page.locator(selector);
  return (await element.textContent()) || "";
}

/**
 * Assert URL contains path
 */
export async function expectURLContains(page: Page, path: string) {
  await expect(page).toHaveURL(new RegExp(path));
}

/**
 * Retry action until it succeeds
 */
export async function retryUntilSuccess<T>(
  action: () => Promise<T>,
  maxAttempts = 3,
  delayMs = 1000,
): Promise<T> {
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      return await action();
    } catch (error) {
      if (attempt === maxAttempts) {
        throw error;
      }
      await new Promise((resolve) => setTimeout(resolve, delayMs));
    }
  }
  throw new Error("Should not reach here");
}

/**
 * Debug: Log current page state
 */
export async function debugPageState(page: Page) {
  console.log("Current URL:", page.url());
  console.log("Title:", await page.title());
  const cookies = await page.context().cookies();
  console.log("Cookies:", cookies.length);
  const localStorage = await page.evaluate(() => Object.keys(localStorage));
  console.log("LocalStorage keys:", localStorage);
}

File: e2e-tests/utils/test_helpers.py
"""
Test helper utilities
"""

import re
from typing import Optional
from playwright.sync_api import Page, expect


def wait_for_navigation(page: Page, url: Optional[str] = None) -> None:
    """Wait for navigation to complete"""
    if url:
        page.wait_for_url(url)
    else:
        page.wait_for_load_state("networkidle")


def wait_for_api_response(page: Page, url_pattern: str | re.Pattern):
    """Wait for API response"""
    def check_url(response):
        url = response.url
        if isinstance(url_pattern, str):
            return url_pattern in url
        return url_pattern.search(url) is not None

    return page.wait_for_response(check_url)


def is_visible(page: Page, selector: str) -> bool:
    """Check if element is visible"""
    try:
        element = page.locator(selector)
        return element.is_visible(timeout=5000)
    except Exception:
        return False


def wait_for_text(page: Page, text: str, timeout: int = 10000) -> None:
    """Wait for text to appear"""
    page.wait_for_selector(f"text={text}", timeout=timeout)


def fill_by_label(page: Page, label: str, value: str) -> None:
    """Fill form field by label"""
    input_selector = f'label:has-text("{label}") + input, label:has-text("{label}") input'
    page.locator(input_selector).fill(value)


def click_button(page: Page, text: str) -> None:
    """Click button by text"""
    page.click(f'button:has-text("{text}")')


def expect_toast(page: Page, message: str) -> None:
    """Assert toast/notification appears"""
    toast = page.locator(
        f'[role="alert"]:has-text("{message}"), .toast:has-text("{message}")'
    )
    expect(toast).to_be_visible(timeout=5000)


def wait_for_loading_complete(page: Page) -> None:
    """Assert loading state completes"""
    try:
        page.wait_for_selector(
            '.loading, [data-loading="true"], .skeleton',
            state="hidden",
            timeout=10000,
        )
    except Exception:
        # If no loading indicator found, that's fine
        pass


def get_text_content(page: Page, selector: str) -> str:
    """Get text content of element"""
    element = page.locator(selector)
    return element.text_content() or ""


def expect_url_contains(page: Page, path: str) -> None:
    """Assert URL contains path"""
    expect(page).to_have_url(re.compile(path))


def retry_until_success(action, max_attempts: int = 3, delay_ms: int = 1000):
    """Retry action until it succeeds"""
    import time

    for attempt in range(1, max_attempts + 1):
        try:
            return action()
        except Exception as e:
            if attempt == max_attempts:
                raise e
            time.sleep(delay_ms / 1000)


def debug_page_state(page: Page) -> None:
    """Debug: Log current page state"""
    print("Current URL:", page.url)
    print("Title:", page.title())
    cookies = page.context.cookies()
    print("Cookies:", len(cookies))
    local_storage = page.evaluate("() => Object.keys(localStorage)")
    print("LocalStorage keys:", local_storage)
